In [190]:
#import pandas and numpy libraries
import numpy as np
import pandas as pd

#import dataset file from URL
climate_change=pd.read_csv(r"https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv")


In [77]:
#print first few data
climate_change.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

QUESTION 1 - How do you create an identity matrix in python?

In [79]:
np.identity(3) #gives an identity matrix

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

QUESTION 2 - For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [87]:
#using groupby function to help me to calculate the mean across the required years 

#fcost_year = climate_change.groupby(['report_year', 'fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].sum()

#loc label to extract the two given years and the fuel type
#y1994 = fcost_year.loc[[1998, 1994], 'coal']
#y1998 = fcost_year.loc[[1994, 1998], 'coal']

#percentage change will be given by the formula
#percent_change = ((y1998 - y1994)/ y1994) * 100
#percentage_change given

report_year  fuel_type_code_pudl
1994         coal                   0.0
1998         coal                   0.0
Name: fuel_cost_per_unit_burned, dtype: float64

In [81]:
fcost_year=climate_change.groupby(['report_year', 'fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].sum()


In [82]:
fcost_year.head()

report_year  fuel_type_code_pudl
1994         coal                    14984.572
             gas                     10792.496
             nuclear                227983.354
             oil                      9362.194
             other                     116.737
Name: fuel_cost_per_unit_burned, dtype: float64

In [96]:
y1994,y1998 = fcost_year.loc[[1998, 1994], 'coal']

In [97]:
y1994

14984.572000000011

In [98]:
y1998

11902.597000000012

In [99]:
percentage_change=((y1998 - y1994)/ y1994) * 100

In [100]:
percentage_change

-20.56765451826049

QUESTION 3 - Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [103]:
# Correlation of other features  compared to the fuel cost burned

climate_change.corr().loc['fuel_cost_per_unit_burned'].sort_values()

#

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

In [112]:
#The second lowest correlation is

climate_change.corr().loc['fuel_cost_per_unit_burned'].sort_values()[1] 

-0.018535416794825232

In [113]:
#The third lowest is

climate_change.corr().loc['fuel_cost_per_unit_burned'].sort_values()[2] 

-0.010033808428852969

The second and third lowest correlation are fuel_qty_burned and fuel_mmbtu_per_unit

QUESTION 4 - The feature with missing values falls under what category? What missing value imputation technique would you use?

In [104]:
#we have categorical can take on one of a limited, and usually fixed, number of possible values, assigning each individual or other unit of observation to a particular group or nominal category on the basis of some qualitative property.
#and it replaces missing values of a categorical variable by the mode of non-missing cases of that variable.
#The answer is categorical and mode imputation

QUESTION 5 - Which year has the highest average fuel cost per unit delivered?

In [202]:
# groupby to combine the year and the fuel cost delivered
# taking the mean, to get the average cost delivered per year
#  sorting the values from highest to the lowest
climate_change.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

The highest average fuel year is 1997

QUESTION 6 - Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places? 
(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)


In [151]:
#Total = climate_change['fuel_unit'].isnull().sum()

#rows_total = fuel_data['record_id'].count()
#rows_total_fuel_unit = fuel_data['fuel_unit'].count()

#percent= (Total/rows_total) * 100


In [146]:
Feature = ('fuel_unit')
Feature

'fuel_unit'

In [141]:
Total = climate_change['fuel_unit'].isnull().sum()

In [142]:
Total

180

In [147]:
total_value = climate_change['record_id'].count()

In [148]:
total_value1 = climate_change['fuel_unit'].count()

In [158]:
percentage = (null/total_value) * 100

In [159]:
percentage

p=percentage

#percentage i three decimal places
print('%.3f' %p)

0.610


Solution: Feature: fuel_unit, Total: 180, Percent: 0.610

QUESTION 7 - What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [152]:
#skewness for the fuel quantity burned
climate_change['fuel_qty_burned'].skew()

15.851495469109503

In [155]:
#skewness for the fuel quantity burned in two decimal places
a=climate_change['fuel_qty_burned'].skew()
print('%.2f' %a)

15.85


In [160]:
#kurtosis of the fuel quantity burned
climate_change['fuel_qty_burned'].kurt()

651.3694501337732

In [161]:
#kurtosis for the fuel quantity burned in two decimal places
b=climate_change['fuel_qty_burned'].kurt()
print('%.2f' %b)

651.37


QUESTION 8 - What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [163]:
#standard deviation
std_deviation = climate_change['fuel_mmbtu_per_unit'].std()

In [164]:
#standard deviation in two decimal places
s = std_deviation = climate_change['fuel_mmbtu_per_unit'].std()
print('%.2f' %s)

10.60


In [167]:
#percentile
climate_change['fuel_mmbtu_per_unit'].quantile(q=0.75, interpolation='linear')

17.006

In [168]:
#percentile deviation in two decimal places

p= climate_change['fuel_mmbtu_per_unit'].quantile(q=0.75, interpolation='linear')
print('%.2f' %p)

17.01


QUESTION 9 - Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [170]:
#grouping the different fuel type
climate_change.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean().sort_values()

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

In [173]:
#lowest average fuel cost per unit burned is
print('gas')

gas


QUESTION 10 - A = [1,2,3,4,5,6], B = [13, 21, 34], which operation gives A_B =  [1,2,3,4,5,6,13, 21, 34]

In [199]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]
A.extend(B)

In [201]:
#A_B is given by new value of A
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]